### ISSUE TITLE
Error uploading sheel script: http response error: 401 - invalid content type

### ISSUE THREAD

**Issue by amitky-vmw** [taliesins/terraform-provider-hyperv#55]
### Terraform Version
Terraform v0.14.8
+ provider registry.terraform.io/taliesins/hyperv v1.0.1

### Affected Resource(s)
No resources are being created, terraform is failing on Apply

### Debug Output

![Screenshot 2021-03-20 at 8 43 45 PM](https://user-images.githubusercontent.com/77483290/111874897-222b0100-89bd-11eb-8e45-203785a97356.png)

**Comment by AbdouDIEYE** [comment]
I got the same error as you but by changing the variable as follows, the winrm connection went through

 terraform {
  required_providers {
    hyperv = {
      version = "1.0.3"
      source = "registry.terraform.io/taliesins/hyperv"
    }
  }
}


provider "hyperv" {
  user     = "administrator"
  password = "******"
  host     = "172.30.***.***"
  port     = "5985"
  https           = false
  insecure        = true
  use_ntlm        = false
  script_path     = "C:/Temp/terraform_%RAND%.cmd"
  timeout         = "30s"
}

**Comment by RBadeloe** [comment]
this solved my similar issue. Thnx

**Comment by paultempleusan** [comment]
I'm having the same issue as the @amitky-vmw above. Tried using the variables that @AbdouDIEYE has listed above, but still receiving the same 401 - invalid content error. Using Server 2019 (1809) and Terraform 0.15.

```
[nor2k19test1]: PS C:\Terraform\Config\tf-hpv-provider-poc> terraform apply tf-hpv-poc
hyperv_network_switch.dmz_network_switch: Creating...
hyperv_vhd.usanctrl_tf_01_vhd: Creating...
terraform :
    + CategoryInfo          : NotSpecified: (â•·:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

Error: error uploading shell script: http response error: 401 - invalid content type

  on main.tf line 26, in resource "hyperv_network_switch" "dmz_network_switch":
  26: resource "hyperv_network_switch" "dmz_network_switch" {

Error: error uploading shell script: http response error: 401 - invalid content type

   on main.tf line 30, in resource "hyperv_vhd" "usanctrl_tf_01_vhd":
   30: resource "hyperv_vhd" "usanctrl_tf_01_vhd" {
```

Below is my main.tf contents... Any assistance is greatly appreciated :) 

```
#####################################
## Terraform Hyper-V Provider POC  ##
#####################################

terraform {
  required_providers {
    hyperv = {
      version = "1.0.3"
      source  = "registry.terraform.io/taliesins/hyperv"
    }
  }
}

provider "hyperv" {
  user     = "local.username"
  password = "local.password"
  host     = "172.22..."
  port     = "5985"
  https    = false
  insecure = true
  use_ntlm = false
  script_path = "C:/Temp/terraform_%RAND%.cmd"
  timeout = "30s"
}

resource "hyperv_network_switch" "dmz_network_switch" {
  name = "dmz"
}

resource "hyperv_vhd" "usanctrl_tf_01_vhd" {
  path = "c:\\vhdx\\usanctrl-tf-01.vhdx" #Needs to be absolute path
  size = 10737418240                     #10GB
}

resource "hyperv_machine_instance" "usanctrl_tf_01" {
  name                   = "usanctrl_tf_01"
  generation             = 1
  processor_count        = 2
  static_memory          = true
  memory_startup_bytes   = 536870912 #512MB
  wait_for_state_timeout = 10
  wait_for_ips_timeout   = 10

  vm_processor {
    expose_virtualization_extensions = true
  }

  network_adaptors {
    name         = "wan"
    switch_name  = hyperv_network_switch.dmz_network_switch.name
    wait_for_ips = false
  }

  hard_disk_drives {
    controller_type     = "Ide"
    path                = hyperv_vhd.usanctrl_tf_01_vhd.path
    controller_number   = 0
    controller_location = 0
  }

  dvd_drives {
    controller_number   = 0
    controller_location = 1
    path                = "C:\\ISO\\Win_Server_2019_STD_CORE_1809.1_64Bit_EN.ISO"
  }
}

```

**Comment by AbdouDIEYE** [comment]
@paultempleusan  Try this it to work for me too. Make sure winrm is configured correctly that port 5986 is open.

provider "hyperv" {
  user     = "administrator"
  password = "*****************"
  host     = "192.xxx.xx.xx"
  port     = "5986"
  https           = true
  insecure        = true
  use_ntlm        = false
  .
  .
  . 
}

**Comment by paultempleusan** [comment]
@AbdouDIEYE Thanks for the response! So I tried the following, as you posted (FYI, I'm running this Terraform locally on the Hyper-V host itself, tried using 127.0.0.1 as well as the actual IP Address, but getting the same 401 errors from above. Reviewed that Port 5986 / 5985 are both open with the command "netstat -nao | more" and it is. Also tried going back to Terraform version 14.10 instead of 15, but same results. I think this is something to do with Server 2019 support, as if I turn on debugging in the PowerShell environment variables.... 

```
$ENV:WINRMCP_DEBUG=1
$ENV:TF_LOG="DEBUG"
```

I see all kinds of PowerShell script errors, including some errors regarding 7-Zip being needed for the creation of the .VHDX. Which I went ahead and installed, but still receiving the 401 error.

```
provider "hyperv" {
  user        = "local.username"
  password    = "local.password"
  host        = "172.XXX.XXX.XXX"
  port        = "5986"
  https       = true
  insecure    = true
  use_ntlm    = false
  script_path = "C:/Temp/terraform_%RAND%.cmd"
  timeout  
  = "30s"
}
``` 
Attached is the debug results when I run "terraform apply "plan name". Its rather long, so put it in its own file. 

[terraform-hyperv-debug.txt](https://github.com/taliesins/terraform-provider-hyperv/files/6378431/terraform-hyperv-debug.txt)

Also here's my winrm config results when running "winrm get winrm/config"

```
Config
    MaxEnvelopeSizekb = 500
    MaxTimeoutms = 1800000
    MaxBatchItems = 32000
    MaxProviderRequests = 4294967295
    Client
        NetworkDelayms = 5000
        URLPrefix = wsman
        AllowUnencrypted = false
        Auth
            Basic = true
            Digest = true
            Kerberos = true
            Negotiate = true
            Certificate = true
            CredSSP = false
        DefaultPorts
            HTTP = 5985
            HTTPS = 5986
        TrustedHosts = *
    Service
        RootSDDL = O:NSG:BAD:P(A;;GA;;;BA)(A;;GR;;;IU)S:P(AU;FA;GA;;;WD)(AU;SA;GXGW;;;WD)
        MaxConcurrentOperations = 4294967295
        MaxConcurrentOperationsPerUser = 1500
        EnumerationTimeoutms = 240000
        MaxConnections = 300
        MaxPacketRetrievalTimeSeconds = 120
        AllowUnencrypted = true
        Auth
            Basic = false
            Kerberos = true
            Negotiate = true
            Certificate = false
            CredSSP = false
            CbtHardeningLevel = Relaxed
        DefaultPorts
            HTTP = 5985
            HTTPS = 5986
        IPv4Filter = *
        IPv6Filter = *
        EnableCompatibilityHttpListener = false
        EnableCompatibilityHttpsListener = false
        CertificateThumbprint
        AllowRemoteAccess = true
    Winrs
        AllowRemoteShellAccess = true
        IdleTimeout = 7200000
        MaxConcurrentUsers = 2147483647
        MaxShellRunTime = 2147483647
        MaxProcessesPerShell = 2147483647
        MaxMemoryPerShellMB = 1024
        MaxShellsPerUser = 2147483647
```

Again, thanks for any time you can spare on the assistance. I'll be sure to post a fix if I find it, as I'm sure others may be experiencing similar results.

**Comment by malikumaliku** [comment]
I have the same issue...

The question is the 
    Service
        Auth
            Basic = false
            Kerberos = true
            Negotiate = true

If on the server part basic = true... I get it working...
But with Basic=false even with use_ntlm=true in the hyperv porvider settings.... I get the same error

And I cannot keep Basic=True due to GPO ....
        Auth
            Basic = false [Source="GPO"]

**Comment by honzajscz** [comment]
My problem was that I have changed the password 🔑 of the hyper-v user (_var.password_ below) and totally forgot about it.  🤦
```
provider "hyperv" {
  user = var.user
  password = var.password
  ...
}
```

**Comment by rismoney** [comment]
Are there any other solutions for Win2019+ for this?  Is this provider still under development?

**Comment by marcoshaw** [comment]
Windows 10 Pro - I got it working after a bit.

You can get this error with an invalid user account.  For example, I didn't want to play around with my main Live account, and I don't quite know how Administrator works on the new OS...  So I created a Administrator2 account as a local Administrator.

I did quite a few things, like enabling a listener for HTTP/5985 (winrm enumerate winrm/config/listener), and I don't quite remember how (I'm sure Googling will tell you), but also:

I think this is part of the solution:
https://devblogs.microsoft.com/powershell/compromising-yourself-with-winrms-allowunencrypted-true/

Now, the above doesn't include in our case:
winrm set winrm/config/client @{AllowUnencrypted="true"}

This is old and I don't know if it's resolved:
https://github.com/hashicorp/terraform/issues/2918

Basically it says at one spot that the Terraform WinRM Go library was configured for Basic only.  I don't know if that's fixed or not.

**Comment by marcoshaw** [comment]
Note there's quite a long script on the main page that shows how to configure WinRM HTTPS.  I didn't try it, I guess if Basic authentication is still required, the communication could at least be secured using HTTPS even though that means NTLM/Kerberos won't be used to pass the authentication creds/token?

**Comment by rismoney** [comment]
All my interfaces are set to private or domain.

I have tried setting winrm to allow unencrypted and all the settings in the 'compromising yourself'
As that didn't work, I then tried to set the configs to use the certs, and went through the README 'quite a long script.  I even added the devcertca to my endpoints Trusted Root Authority CA folder, and my hyperv node.
I even tried issuing a server certificate from my internal CA.  None of these work.
I am on a Win2019 core, fully patched vanilla install with hyperv roll enabled, and a win10 endpoint.

The logs don't provide much, and i am just trying to making the most minimalistic of configs as a test:

```
  # hyperv_network_switch.dmz_network_switch will be created
  + resource "hyperv_network_switch" "dmz_network_switch" {
      + allow_management_os                     = true
      + default_flow_minimum_bandwidth_absolute = 0
      + default_flow_minimum_bandwidth_weight   = 0
      + default_queue_vmmq_enabled              = false
      + default_queue_vmmq_queue_pairs          = 16
      + default_queue_vrss_enabled              = false
      + enable_embedded_teaming                 = false
      + enable_iov                              = false
      + enable_packet_direct                    = false
      + id                                      = (known after apply)
      + minimum_bandwidth_mode                  = "None"
      + name                                    = "dmz"
      + switch_type                             = "Internal"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes
2022/02/13 23:22:32 [INFO] backend/local: apply calling Apply

2022/02/13 23:22:32 [INFO] terraform: building graph: GraphTypeApply
2022/02/13 23:22:32 [DEBUG] Resource state not found for node "hyperv_network_switch.dmz_network_switch", instance hyperv_network_switch.dmz_network_switch
2022/02/13 23:22:32 [DEBUG] ProviderTransformer: "hyperv_network_switch.dmz_network_switch (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/taliesins/hyperv"]
2022/02/13 23:22:32 [DEBUG] ProviderTransformer: "hyperv_network_switch.dmz_network_switch" (*terraform.NodeApplyableResourceInstance) needs provider["registry.terraform.io/taliesins/hyperv"]
2022/02/13 23:22:32 [DEBUG] ReferenceTransformer: "hyperv_network_switch.dmz_network_switch (expand)" references: []
2022/02/13 23:22:32 [DEBUG] ReferenceTransformer: "hyperv_network_switch.dmz_network_switch" references: []
2022/02/13 23:22:32 [DEBUG] ReferenceTransformer: "var.name" references: []
2022/02/13 23:22:32 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/taliesins/hyperv\"]" references: []
2022/02/13 23:22:32 [DEBUG] Starting graph walk: walkApply
2022-02-13T23:22:32.612-0500 [INFO]  plugin: configuring client automatic mTLS
2022-02-13T23:22:32.636-0500 [DEBUG] plugin: starting plugin: path=.terraform/plugins/registry.terraform.io/taliesins/hyperv/1.0.3/windows_amd64/terraform-provider-hyperv_1.0.3.exe args=[.terraform/plugins/registry.terraform.io/taliesins/hyperv/1.0.3/windows_amd64/terraform-provider-hyperv_1.0.3.exe]
2022-02-13T23:22:32.639-0500 [DEBUG] plugin: plugin started: path=.terraform/plugins/registry.terraform.io/taliesins/hyperv/1.0.3/windows_amd64/terraform-provider-hyperv_1.0.3.exe pid=6796
2022-02-13T23:22:32.639-0500 [DEBUG] plugin: waiting for RPC address: path=.terraform/plugins/registry.terraform.io/taliesins/hyperv/1.0.3/windows_amd64/terraform-provider-hyperv_1.0.3.exe
2022-02-13T23:22:32.657-0500 [INFO]  plugin.terraform-provider-hyperv_1.0.3.exe: configuring server automatic mTLS: timestamp=2022-02-13T23:22:32.657-0500
2022-02-13T23:22:32.692-0500 [DEBUG] plugin: using plugin: version=5
2022-02-13T23:22:32.692-0500 [DEBUG] plugin.terraform-provider-hyperv_1.0.3.exe: plugin address: address=127.0.0.1:10000 network=tcp timestamp=2022-02-13T23:22:32.692-0500
2022-02-13T23:22:32.740-0500 [INFO]  plugin.terraform-provider-hyperv_1.0.3.exe: 2022/02/13 23:22:32 [INFO][hyperv] HyperV HypervClient configured for HyperV API operations using:
  Host: NYCOWSHYP-P22
  Port: 5986
  User: testing
  Password: true
  HTTPS: true
  Insecure: false
  NTLM: true
  TLSServerName:
  CACert: false
  Cert: false
  Key: false
  ScriptPath: C:/Temp/terraform_%RAND%.cmd
  Timeout: 30s: timestamp=2022-02-13T23:22:32.740-0500
2022/02/13 23:22:32 [WARN] Provider "registry.terraform.io/taliesins/hyperv" produced an invalid plan for hyperv_network_switch.dmz_network_switch, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .allow_management_os: planned value cty.True does not match config value cty.NullVal(cty.Bool)
      - .enable_embedded_teaming: planned value cty.False does not match config value cty.NullVal(cty.Bool)
      - .default_queue_vmmq_enabled: planned value cty.False does not match config value cty.NullVal(cty.Bool)
h      - .default_queue_vrss_enabled: planned value cty.False does not match config value cty.NullVal(cty.Bool)
y      - .enable_iov: planned value cty.False does not match config value cty.NullVal(cty.Bool)
perv_network_swivalue cty.NullVal(cty.Bool)
tc      - .minimum_bandwidth_mode: planned value cty.StringVal("None") does not match config value cty.NullVal(cty.String)
h.dmz_netw      - .default_flow_minimum_bandwidth_weight: planned value cty.NumberIntVal(0) does not match config value cty.NullVal(cty.Number)
      - .default_queue_vmmq_queue_pairs: planned value cty.NumberIntVal(16) does not match config value cty.NullVal(cty.Number)
ork_sw      - .switch_type: planned value cty.StringVal("Internal") does not match config value cty.NullVal(cty.String)
itch: Creating...      - .default_flow_minimum_bandwidth_absolute: planned value cty.NumberIntVal(0) does not match config value cty.NullVal(cty.Number)
2022/02/13 23:22:32 [DEBUG] EvalApply: ProviderMeta config value set
2022/02/13 23:22:32 [DEBUG] hyperv_network_switch.dmz_network_switch: applying the planned Create change
2022-02-13T23:22:32.748-0500 [INFO]  plugin.terraform-provider-hyperv_1.0.3.exe: 2022/02/13 23:22:32 [DEBUG] Running fire and forget script:


$ErrorActionPreference = 'Stop'
Import-Module Hyper-V
$vmSwitch = '{"Name":"dmz","Notes":"","AllowManagementOS":true,"EmbeddedTeamingEnabled":false,"IovEnabled":false,"PacketDirectEnabled":false,"BandwidthReservationMode":3,"SwitchType":1,"NetAdapterNames":[],"DefaultFlowMinimumBandwidthAbsolute":0,"DefaultFlowMinimumBandwidthWeight":0,"DefaultQueueVmmqEnabled":false,"DefaultQueueVmmqQueuePairs":16,"DefaultQueueVrssEnabled":false}' | ConvertFrom-Json
$minimumBandwidthMode = [Microsoft.HyperV.PowerShell.VMSwitchBandwidthMode]$vmSwitch.BandwidthReservationMode
$switchType = [Microsoft.HyperV.PowerShell.VMSwitchType]$vmSwitch.SwitchType
$NetAdapterNames = @($vmSwitch.NetAdapterNames)
#when EnablePacketDirect=true it seems to throw an exception if EnableIov=true or EnableEmbeddedTeaming=true

$switchObject = Get-VMSwitch -Name "$($vmSwitch.Name)*" | ?{$_.Name -eq $vmSwitch.Name}

if ($switchObject){
        throw "Switch already exists - $($vmSwitch.Name)"
}

$NewVmSwitchArgs = @{}
$NewVmSwitchArgs.Name=$vmSwitch.Name
$NewVmSwitchArgs.MinimumBandwidthMode=$minimumBandwidthMode
$NewVmSwitchArgs.EnableEmbeddedTeaming=$vmSwitch.EmbeddedTeamingEnabled
$NewVmSwitchArgs.EnableIov=$vmSwitch.IovEnabled
$NewVmSwitchArgs.EnablePacketDirect=$vmSwitch.PacketDirectEnabled

tAdapterNames) {
        $NewVmSwitchArgs.AllowManagementOS=$vmSwitch.AllowManagementOS
        $NewVmSwitchArgs.NetAdapterName=$NetAdapterNames
} else {
        $NewVmSwitchArgs.SwitchType=$switchType
        #not used unless interface is specified
        #-AllowManagementOS $vmSwitch.AllowManagementOS
}
New-VMSwitch @NewVmSwitchArgs

$switchObject = Get-VMSwitch -Name "$($vmSwitch.Name)" | ?{$_.Name -eq $vmSwitch.Name}

if (!$switchObject){
        throw "Switch does not exist - $($vmSwitch.Name)"
}

$SetVmSwitchArgs = @{}
$SetVmSwitchArgs.Name=$vmSwitch.Name
$SetVmSwitchArgs.Notes=$vmSwitch.Notes
if (($minimumBandwidthMode -eq [Microsoft.HyperV.PowerShell.VMSwitchBandwidthMode]::Absolute) -and $switchObject.DefaultFlowMinimumBandwidthAbsolute -ne $vmSwitch.DefaultFlowMinimumBandwidthAbsolute) {
        $SetVmSwitchArgs.DefaultFlowMinimumBandwidthAbsolute=$vmSwitch.DefaultFlowMinimumBandwidthAbsolute
}
if ((($minimumBandwidthMode -eq [Microsoft.HyperV.PowerShell.VMSwitchBandwidthMode]::Weight) -or (($minimumBandwidthMode -eq [Microsoft.HyperV.PowerShell.VMSwitchBandwidthMode]::Default) -and (-not ($vmSwitch.IovEnabled)))) -and $switchObject.DefaultFlowMinimumBandwidthWeight -ne $vmSwitch.DefaultFlowMinimumBandwidthWeight) {
        $SetVmSwitchArgs.DefaultFlowMinimumBandwidthWeight=$vmSwitch.DefaultFlowMinimumBandwidthWeight
}
$SetVmSwitchArgs.DefaultQueueVmmqEnabled=$vmSwitch.DefaultQueueVmmqEnabled
$SetVmSwitchArgs.DefaultQueueVmmqQueuePairs=$vmSwitch.DefaultQueueVmmqQueuePairs
$SetVmSwitchArgs.DefaultQueueVrssEnabled=$vmSwitch.DefaultQueueVrssEnabled

Set-VMSwitch @SetVmSwitchArgs: timestamp=2022-02-13T23:22:32.748-0500
2022-02-13T23:22:32.751-0500 [INFO]  plugin.terraform-provider-hyperv_1.0.3.exe: 2022/02/13 23:22:32 [DEBUG] Uploading shell wrapper for command from [C:\Users\AD-RSI~1\AppData\Local\Temp\shell-terraform-255MZrlnfgrHauesrhdPxtVQAkF.ps1998041371] to [$env:TEMP\shell-terraform-255MZrlnfgrHauesrhdPxtVQAkF.ps1]: timestamp=2022-02-13T23:22:32.750-0500
2022/02/13 23:22:32 [DEBUG] hyperv_network_switch.dmz_network_switch: apply errored, but we're indicating that via the Error pointer rather than returning it: error uploading shell script: http response error: 401 - invalid content type
2022/02/13 23:22:32 [ERROR] eval: *terraform.EvalApplyPost, err: error uploading shell script: http response error: 401 - invalid content type
2022/02/13 23:22:32 [ERROR] eval: *terraform.EvalSequence, err: error uploading shell script: http response error: 401 - invalid content type

Error: error uploading shell script: http response error: 401 - invalid content type2022-02-13T23:22:32.798-0500 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
```

**Comment by marcoshaw** [comment]
> All my interfaces are set to private or domain.
> 
> I have tried setting winrm to allow unencrypted and all the settings in the 'compromising yourself' As that didn't work, I then tried to set the configs to use the certs, and went through the README 'quite a long script. I even added the devcertca to my endpoints Trusted Root Authority CA folder, and my hyperv node. I even tried issuing a server certificate from my internal CA. None of these work. I am on a Win2019 core, fully patched vanilla install with hyperv roll enabled, and a win10 endpoint.
> 
> The logs don't provide much, and i am just trying to making the most minimalistic of configs as a test:
> 

So you have:
-Windows 10 -> Windows 2019 Core
-Either or both domain joined?  Same domain?
-You've tripled checked the username/password and are using an Administrator account to connect to Hyper-V?

**Comment by rismoney** [comment]
I have it working now using https.   I got it working using the script from the readme.  need to go through it more, because if I am going this route,. i want to use CA based certs, but I should be good.  Thank you!

**Comment by taliesins** [comment]
@marcoshaw and @rismoney hey guys, seems like its trouble again with setting up WinRM ;>

For guys in the future, please read the readme if you have trouble setting up WinRM.

Please note that the provider supports basic, ntlm and now also kerberos.
